File: logitstic_regression/go/lr.go
package main

import (
    "fmt"
    "math"
)

type LogisticRegression struct {
    weights []float64
    bias float64
}

func (model *LogisticRegression) Train(features []float64, label int) {
    prediction := sigmoid(dot(features, model.weights) + model.bias)
    error := float64(label) - prediction

    for i, feature := range features {
        model.weights[i] += error * feature
    }
    model.bias += error
}

func (model *LogisticRegression) Predict(features []float64) int {
    return int(sigmoid(dot(features, model.weights) + model.bias) > 0.5)
}

func sigmoid(x float64) float64 {
    return 1.0 / (1.0 + math.Exp(-x))
}

func dot(features []float64, weights []float64) float64 {
    sum := 0.0
    for i, feature := range features {
        sum += feature * weights[i]
    }
    return sum
}

func main() {
    features := []float64{1.0, 2.0}
    label := 1

    model := LogisticRegression{
        weights: make([]float64, len(features)),
        bias: 0.0,
    }

    for i := 0; i < 100; i++ {
        model.Train(features, label)
    }

    newFeatures := []float64{3.0, 4.0}
    newLabel := model.Predict(newFeatures)

    fmt.Println("The predicted label is", newLabel)
}

File: logitstic_regression/julia/lr.jl
struct LogisticRegression
    weights::Array{Float64,1}
    bias::Float64
  end
  
  function train(model::LogisticRegression, features::Array{Float64,1}, label::Int64)
    prediction = sigmoid(dot(model.weights, features) + model.bias)
    error = label - prediction
  
    for i in 1:length(features)
      model.weights[i] += error * features[i]
    end
    model.bias += error
  end
  
  function sigmoid(x::Float64)
    return 1.0 / (1.0 + exp(-x))
  end
  
  function dot(x::Array{Float64,1}, y::Array{Float64,1})
    sum = 0.0
  
    for i in 1:length(x)
      sum += x[i] * y[i]
    end
  
    return sum
  end
  
  function predict(model::LogisticRegression, features::Array{Float64,1})
    prediction = sigmoid(dot(model.weights, features) + model.bias)
    return prediction > 0.5
  end
  
  function main()
    features = [1.0, 2.0]
    label = 1
  
    model = LogisticRegression(Array(Float64, 2))
  
    for i in 1:100
      train(model, features, label)
    end
  
    new_features = [3.0, 4.0]
    new_label = predict(model, new_features)
  
    println("The predicted label is $(new_label)")
  end
  
  main()
  